### PR TITLE
Heap corruption check

### PR DESCRIPTION
**Pull request by juj**
Update the error message of little endian check to avoid referring to typed arrays 2 mode, which is not relevant anymore. Also, only write 0x000000FF to address 0 in the beginning of the main thread so that the check has more power to detect memory corruption when pthread workers are spawned.

### CODE REVIEW

File: src/preamble.js [hunk 1]
```diff
@@ -1422,8 +1422,14 @@ function setF64(ptr, value) {
...
-HEAP32[0] = 255;
-if (HEAPU8[0] !== 255 || HEAPU8[3] !== 0) throw 'Typed arrays 2 must be run on a little-endian system';
+#if USE_PTHREADS
+if (!ENVIRONMENT_IS_PTHREAD) {
+#endif
+  HEAP32[0] = 255;
+#if USE_PTHREADS
+}
+#endif
+if (HEAPU8[0] !== 255 || HEAPU8[3] !== 0) throw 'Runtime error: either the current system is not little-endian, or it has corrupted its heap memory area (address zero)!';
```
Comment (kripken) [inline src/preamble.js]: I'd rather we split the checks: check for endianness, and show an error if it fails. And separately check for corruption and show an error for that if that fails. The reason is that this combined error message means, if someone sees it, they need to figure out which of the 2 occurred, and the second is potentially very confusing.

**Comment by pipcet** [comment]
Sorry for barging in, but I do wonder whether 255 is a good word to leave stored at address 0. It took me a while to figure out that that meant I was deref'ing a null pointer somewhere, so maybe we should use a more magic constant that people can find on the web?

**Comment by kripken** [comment]
Sounds like a good idea. How about "emsc" in hex? (`0x656d7363`)

**Comment by pipcet** [comment]
I'd prefer 0x63736d65, just in case someone manages to print it as a string (which the library would usually catch, anyway); anything unaligned and reasonably unique would work, though. I think Emscripten data currently starts at address 8 so there's a NUL character following that, as well.

**Comment by kripken** [comment]
Yeah, I forgot about little-endian ;)

**Comment by juj** [comment]
Works for me, updated. How about now?

File: src/preamble.js [hunk 2]
```diff
@@ -1425,11 +1427,13 @@ function setF64(ptr, value) {
...
 #endif
-  HEAP32[0] = 255;
+  HEAP32[0] = 0x63736d65; /* 'emsc' */
 #if USE_PTHREADS
 }
 #endif
-if (HEAPU8[0] !== 255 || HEAPU8[3] !== 0) throw 'Runtime error: either the current system is not little-endian, or it has corrupted its heap memory area (address zero)!';
+HEAP16[1] = 0x6373;
+if (HEAPU8[2] !== 0x73 || HEAPU8[3] !== 0x63) throw 'Runtime error: expected the system to be little-endian!';
+if (HEAP32[0] !== 0x63736d65) throw 'Runtime error: The application has corrupted its heap memory area (address zero)!';
```
Comment (kripken) [inline src/preamble.js]: could this line be moved 2 higher, before the 16-bit assign? that way we don't ignore corruption of the area the write writes to?

Comment (juj) [inline src/preamble.js]: Sure, updated. How's that?

File: src/preamble.js [hunk 3]
```diff
@@ -1422,8 +1424,16 @@ function setF64(ptr, value) {
...
-HEAP32[0] = 255;
-if (HEAPU8[0] !== 255 || HEAPU8[3] !== 0) throw 'Typed arrays 2 must be run on a little-endian system';
+#if USE_PTHREADS
+if (!ENVIRONMENT_IS_PTHREAD) {
+#endif
+  HEAP32[0] = 0x63736d65; /* 'emsc' */
+#if USE_PTHREADS
+}
+#endif
+if (HEAP32[0] !== 0x63736d65) throw 'Runtime error: The application has corrupted its heap memory area (address zero)!';
```
Comment (kripken) [inline src/preamble.js]: this check might as well be in an else for the above if, right? it's relevant only for a pthread. lgtm with that.

Comment (juj) [inline src/preamble.js]: yeah, makes sense. Thanks!

**Pull request merged**